### PR TITLE
create order

### PR DESCRIPTION
**Pull request by tranchien2002**

### CODE REVIEW

File: app/models/cart.rb [hunk 1]
```diff
@@ -15,10 +15,15 @@ class Cart < ApplicationRecord
     where.not(session_id: nil)
   }
 
+  scope :cart_order, -> {
+    select("quantity, food_id, total_money")
+  }
+
   def Cart.total
     carts = Cart.all
+    total = 0
```
Comment (chuanhtuan) [inline app/models/cart.rb]: array.map(&:to_i).reduce(0, :+)

File: app/models/cart.rb [hunk 2]
```diff
@@ -15,10 +15,15 @@ class Cart < ApplicationRecord
     where.not(session_id: nil)
   }
 
+  scope :cart_order, -> {
+    select("quantity, food_id, total_money")
+  }
+
   def Cart.total
     carts = Cart.all
```
Comment (chuanhtuan) [inline app/models/cart.rb]: lấy theo từng user

File: app/views/orders/new.html.erb [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    <div id="order-body">
+        <div id="order-customer">
+          <i class="fa fa-check" style="color: #9CA643"></i>
+        <b><%= current_user.name %></b><br>
+        <span><%= t "views.orders.new.address" %> : <%= current_user.address %></span><br>
+        <span><%= t "views.orders.new.phone" %> : </span> <%= current_user.phone_number %>
+        </div>
+        <div id="order-address">
+          <i class="fa fa-map-marker"></i> <span><b><%= t "views.orders.new.shipping_address" %></b></span>
+          <%= form_for @order, url: {action: "create"} do |f| %>
```
Comment (chuanhtuan) [inline app/views/orders/new.html.erb]: action create ???

File: app/views/orders/new.html.erb [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+            <%= f.text_area :address, size: "68x4", id:"order-customer", placeholder: t("views.orders.new.example") %><br>
+            <%= f.submit t("views.orders.new.order"), class:"btn btn-success" %>
+          <% end %>
+        </div>
+
+      </div>
+  </div>
+  <div class="col-sm-3">
+    <div id="order-head">
+      <b><%= t "views.orders.new.info_order" %> ( <%=Cart.count%> <%= t "views.orders.new.dish" %> )</b>
```
Comment (chuanhtuan) [inline app/views/orders/new.html.erb]: count theo từng user

File: app/controllers/orders_controller.rb [hunk 5]
```diff
@@ -22,6 +22,38 @@ def show
...
+      @current_user.carts.delete_all
+      redirect_to root_url
+      flash[:success] = t "controllers.orders.create.success"
+    else
+      redirect_to request.referrer || root_url
+      flash[:info] = t "controllers.orders.create.lost_info"
+    end
+  end
+
+  private
```
Comment (tuanlinhtl17) [inline app/controllers/orders_controller.rb]: add strong params

File: app/controllers/orders_controller.rb [hunk 6]
```diff
@@ -22,6 +22,38 @@ def show
...
+
+  def create
+    carts = Cart.cart_order
+    order_details = to_hash carts
+    if params[:order][:address] != ""
+      address = params[:order][:address]
+    else
+      address = current_user.address
+    end
+    order = Order.new(
```
Comment (tuanlinhtl17) [inline app/controllers/orders_controller.rb]: order = Order.new order_params

File: app/controllers/orders_controller.rb [hunk 7]
```diff
@@ -22,6 +22,38 @@ def show
...
+        customer_id: current_user.id,
+        status: Settings.status.new_order,
+        address: address,
+        total_money: Cart.total(@current_user.id),
+        order_details_attributes: order_details
+    )
+    if order.save
+      @current_user.carts.delete_all
```
Comment (tuanlinhtl17) [inline app/controllers/orders_controller.rb]: remove @

File: app/controllers/carts_controller.rb [hunk 8]
```diff
@@ -52,8 +54,10 @@ def destroy
 
   private
   def update_cart
-    Cart.all.each do |c|
-      c.update_attributes(total_money: c.quantity * c.food.price)
+    if logged_in?
+      Cart.cart_current(@current_user.id).each do |c|
```
Comment (tuanlinhtl17) [inline app/controllers/carts_controller.rb]: remove @

File: app/assets/javascripts/cart.js [hunk 9]
```diff
@@ -8,9 +8,9 @@ function charged(){
     price = $(".price");
     quantity = $(".quantity");
     for (var i = 0; i < price.length; i++){
-        totalPrice =totalPrice +  parseFloat(getPrice.exec(price[i].textContent)[0]) * parseFloat(quantity[i].value) + " $"
+        totalPrice =totalPrice +  parseFloat(getPrice.exec(price[i].textContent)[0]) * parseFloat(quantity[i].value)
```
Comment (tuanlinhtl17) [inline app/assets/javascripts/cart.js]: use +=

File: app/controllers/orders_controller.rb [hunk 10]
```diff
@@ -22,6 +22,38 @@ def show
...
+    if params[:order][:address] != ""
+      address = params[:order][:address]
+    else
+      address = current_user.address
+    end
+    order = Order.new(
+        customer_id: current_user.id,
+        status: Settings.status.new_order,
+        address: address,
+        total_money: Cart.total(@current_user.id),
```
Comment (tuanlinhtl17) [inline app/controllers/orders_controller.rb]: remove @

File: app/views/orders/new.html.erb [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+          <% @orders.each do |o| %>
+           <tr>
+            <td><%= o.food.name %></td>
+            <td><%= o.quantity %></td>
+            <td><%= o.total_money %> $</td>
+           </tr>
+          <% end %>
+          <tr>
+            <td><b><%= t "views.orders.new.total" %> :</b></td>
+            <td><%= Cart.total @current_user.id %> $</td>
```
Comment (tuanlinhtl17) [inline app/views/orders/new.html.erb]: remove @

File: app/views/orders/new.html.erb [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+            <%= f.text_area :address, size: "68x4", id:"order-customer", placeholder: t("views.orders.new.example") %><br>
+            <%= f.submit t("views.orders.new.order"), class:"btn btn-success" %>
+          <% end %>
+        </div>
+
+      </div>
+  </div>
+  <div class="col-sm-3">
+    <div id="order-head">
+      <b><%= t "views.orders.new.info_order" %> ( <%=@current_user.carts.count%> <%= t "views.orders.new.dish" %> )</b>
```
Comment (tuanlinhtl17) [inline app/views/orders/new.html.erb]: remove @

File: app/models/cart.rb [hunk 13]
```diff
@@ -11,16 +11,18 @@ class Cart < ApplicationRecord
...
+    select("quantity, food_id, total_money")
   }
 
-  def Cart.total
-    carts = Cart.all
-    carts.each do |c|
-      total = total +  c.quantity * c.price
-    end
-    return total
+  scope :cart_current, -> user_id{
```
Comment (tuanlinhtl17) [inline app/models/cart.rb]: use current_user.carts

File: app/models/cart.rb [hunk 14]
```diff
@@ -11,16 +11,18 @@ class Cart < ApplicationRecord
...
+    where(user_id: user_id)
+  }
+
+
+  def Cart.total id
+    carts = Cart.cart_current id
```
Comment (tuanlinhtl17) [inline app/models/cart.rb]: carts = current_user.carts

File: app/views/orders/new.html.erb (hunk 12 above)
Comment (tuanlinhtl17) [inline app/views/orders/new.html.erb]: @

File: app/controllers/carts_controller.rb (hunk 8 above)
Comment (tuanlinhtl17) [inline app/controllers/carts_controller.rb]: current_user.carts.each

File: app/controllers/carts_controller.rb [hunk 15]
```diff
@@ -52,8 +54,10 @@ def destroy
 
   private
   def update_cart
-    Cart.all.each do |c|
-      c.update_attributes(total_money: c.quantity * c.food.price)
+    if logged_in?
```
Comment (tuanlinhtl17) [inline app/controllers/carts_controller.rb]: && current_user.carts.size != 0

File: app/models/cart.rb [hunk 16]
```diff
@@ -11,16 +11,18 @@ class Cart < ApplicationRecord
     where(session_id: session_id)
   }
 
-  scope :all_cart, -> {
-    where.not(session_id: nil)
+  scope :cart_order, -> {
```
Comment (tuanlinhtl17) [inline app/models/cart.rb]: scope :cart_order, -> customer_id {
     //just get carts with correct customer_id
}

File: app/controllers/carts_controller.rb [hunk 17]
```diff
@@ -52,8 +57,10 @@ def destroy
 
   private
   def update_cart
-    Cart.all.each do |c|
-      c.update_attributes(total_money: c.quantity * c.food.price)
+    if logged_in? && current_user.carts.size != 0
+      current_user.carts.each do |c|
+        c.update_attributes(total_money: c.quantity * c.food.price)
```
Comment (chuanhtuan) [inline app/controllers/carts_controller.rb]: call back function

File: app/assets/javascripts/cart.js [hunk 18]
```diff
@@ -29,15 +29,13 @@ function update_cart(){
...
     $.ajax({
         url: "/carts",
-        type: "put",
-        dataType: "json",
-        data: {cart:{id: ids,quantity: quantitys}},
-        success: function (res) {
-        }
+        type: "patch",
+        dataType: "script",
+        data: {cart:{id: ids,quantity: quantitys}}
```
Comment (chuanhtuan) [inline app/assets/javascripts/cart.js]: OOP

**Pull request merged**